measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+        """Combine Jacobian Q for each measurement to be in one Jacobian Q.
+        Through _split_jacobian we get a list of lists for Jacobian Q,
+        each list contains an Nt*n_parameters elements, which is the sensitivity matrix Q for measurement m
+
+        We aim to convert this list of arrays to Q, a numpy array containing Jacobian matrix of the shape N_total_m * Np
+        Jacobian Q structure:
+        [ \partial y1(t1)/ \partial p1, ..., \partial y1(t1)/ \partial pN,
+        \partial y1(t2)/ \partial p1, ..., \partial y1(t2)/ \partial pN,
+        ...,
+        \partial y1(tN)/ \partial p1, ..., \partial y1(tN)/ \partial pN,
+        \partial y2(t1)/ \partial p1, ..., \partial y2(t1)/ \partial pN,
+        ...,
+        \partial yN(tN)/ \partial p1, ..., \partial yN(tN)/ \partial pN,]
+
+        Arguments
+        ---------
+        :param static_measurement_idx: list of the index for static measurements
+        :param dynamic_measurement_idx: list of the index for dynamic measurements
+
+        Returns
+        -------
+        self.jac: Jacobian information for main class use, a 2D numpy array
+        """
+
+        # get the maximum index from index set
+        # why not sum up static index number and dynamic index number? because they can overlap
+        max_measure_idx = max(max(static_measurement_idx), max(dynamic_measurement_idx))
+        # the total rows of Q should be equal or more than the number of maximum index given by the argument
+        if len(self.jacobian) < max_measure_idx * self.Nt:
+            raise ValueError(
+                "Inconsistent Jacobian matrix shape. Expecting at least "
+                + str(max_measure_idx * self.Nt)
+                + " rows in Jacobian matrix Q."
+            )
+
+        # if one measurement is in both SCM indices and DCM indices, its Jacobian is included twice (overlap is allowed)
+        # compute how many measurements as SCM
+        if static_measurement_idx:
+            static_idx_len = len(static_measurement_idx)
+        # if given None, this variable is given 0 (len(None) gives error message)
+        else:
+            static_idx_len = 0
+
+        # compute how many measurements as DCM
+        if dynamic_measurement_idx:
+            dynamic_idx_len = len(dynamic_measurement_idx)
+        # if given None, this variable is given 0 (len(None) gives error message)
+        else:
+            dynamic_idx_len = 0
+
+        # compute N_total_measure, including both SCMs and DCMs. store for later use
+        self.total_measure_idx = static_idx_len + dynamic_idx_len
+        # all measurements have Nt time points
+        total_measure_len = (static_idx_len + dynamic_idx_len) * self.Nt
+
+        # initialize Jacobian Q as numpy array
+        # here we stack the Jacobian Q according to the orders the user provides SCM and DCM index
+        # jac: Jacobian matrix of shape N_total_measure * Np
+        jac = np.zeros((total_measure_len, self.n_parameters))
```

Comment:
This is a numpy array, not a list of lists. Please update the doc string and comments accordingly.

Reply:
changed